### ISSUE TITLE
"no applicable method for 'predict' when using 'ranger' random forest

### ISSUE THREAD

**Issue by quantumlinguist** [mayer79/kernelshap#58]
I'm trying to calculate SHAP values for a rf with 2300 trees and I keep getting this message. 

Error in kernelshap_one(x = X[i, , drop = FALSE], v1 = v1[i, , drop = FALSE],  : 
  task 1 failed - "no applicable method for 'predict' applied to an object of class "ranger""

The demo with the diamonds dataset works fine so I don't know why it doesn't work with my own data. 

The rf syntax looks like this:

rf= ranger(OverOrdAltiAltFactor~., 
           data= df,
           num.trees = 2300,
           probability = TRUE)

If I dont use "probability= TRUE" I get an error that "predictions must be a vector, matrix, data.frame or 2D array". 

I can't share the data unfortunately.

**Comment by quantumlinguist** [comment]
I will reply to my own post in case this happens to anyone else. I fixed this by adding "parallel_args=list(.packages= "ranger") so the final syntax looks like this:

shaps= kernelshap(rf, sample_big, bg_X= sample_small, parallel=TRUE,
parallel_args=list(.packages= "ranger"))

This is only relevant when parallel= TRUE, according to the documentation of the package.

**Comment by mayer79** [comment]
Ohhh, yes. This is an unfortunate issue with the `predict()` function: It is using `stats::predict()`, so the parallel backend thinks that "stats" package is doing the work. Thus, it does not automatically detect that "ranger" is necessary.

See the discussion in https://github.com/HenrikBengtsson/future/issues/641

Out of curiosity: Do you see a speedup with the parallel backend and ranger?

**Comment by quantumlinguist** [comment]
No, not at all (but i can't really compare because I haven't tried without it). I'm working with a relatively small dataset of 6000 datapoints to fit the random forest and calculating the kernelshap with a subsample of 2000 and bg_X argument is a sample of 200 and I still haven't managed to complete the process. My mac laptop crashed after many hours and now I'm using the server at my job so hopefully that will do the job but it's been going for 1 hour now already.

**Comment by mayer79** [comment]
Tough job, indeed. How many features are in the data?

You can check the necessary time with a subsample of 2 and then multiply with 1000.

**Comment by quantumlinguist** [comment]
Ah nice trick. It would take 11 hours 😩😩😩

**Comment by mayer79** [comment]
2300 deep trees is very suboptimal for SHAP, even for TreeSHAP. Any chance to reduce it to the usual 500? 

Furthermore, I'd opt for decomposing only 500 rows, and or reducing the size of the background data to 50. And maybe no parallel mode (ranger *is* parallelized, although I don't know how well this works at prediction stage).

**Comment by quantumlinguist** [comment]
hey, thank you for all your feedback! 2300 trees was determined through cross-validation. I could check how much difference in accuracy there is with only 500. 

But, I did do something similar to what you suggested (I had done it before you sent your suggestion). So I did 500 rows with 200 rows for the background data. That is fine, it took 1.25 hours. 

I don't quite understand how the background data is used, do you have any suggestions where I can read about this? Thanks!

**Comment by mayer79** [comment]
In Python, you will get a warning if the background data reached 100 rows (or even less). Scott suggest to condense the data via cluster analysis to just very few lines, which is an interesting idea. I don't have a clear opinion yet on the importance of the background data. 

For independent data rows and random forests, the more trees the better. So this parameter is usually not automatically selected by validation but rather by time or storage space constraints. For dependent rows, things anyway get complicated.
